Add the `IllegalMethodCalledDuringTaskExecution` ErrorProne to prevent `getProject()` calls from within a task action (#142)
palantir/gradle-guide: guide/adopting-the-configuration-cache.md

<!-- PreviousNext:START -->
<table><tr>
  <td>Previous: <a href="testing-and-running-your-plugins-locally.md">Testing and running your plugins</a></td>
  <td align="center"><a href="../README.md#table-of-contents">Table of Contents</a></td>
  <td align="right">Next: <a href="diagnosing-build-performance.md">Diagnosing Build Performance</a></td>
</tr></table>
<!-- PreviousNext:END -->

> # Adopting the Configuration Cache

Reviewer: I think this needs a brief overview of how the configuration cache works, ie it saves the input/output properties of each task and loads it back later to avoid the configuration stage, it regards any external process/file read/gradle property etc as an input to configuration time that is evaluated every time it's run to see if the cache is busted or not.

Author: I didn't want to repeat any information from the user guide. Do you think we should explain the fundamentals of the CC here as well? Or maybe we can point people towards that section of the guide?

Author: As discussed in person, this guide should be relatively self-contained, and help people to understand enough to dive into the gradle user guide themselves. 

Let's explain the fundamentals of the Configuration Cache here.

Builds run with the Configuration Cache optionally in Gradle 8, by default in Gradle 9, and necessarily by Gradle 10. Furthermore, your Gradle build has to fulfill [strict requirements]([this section of the Gradle User Guide](https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:requirements)) to be Configuration Cachable. This section aims to be a practical guide for adopting the Configuration Cache in your Gradle projects.

## Finding Configuration Cache problems

To find Configuration Cache problems:

1. Run `./gradlew build --configuration-cache`
2. Find the configuration cache problems in the output. If there are no problems, congratulations! Your build is Configuration Cache compatible.
3. If not, fix the problems. 
4. Run `./gradlew build --configuration-cache` again. In our experience, this uncovers more problems that weren't there in (1). 

The problems you find in step (2) are usually one of two kinds:

1. **external process started `/usr/bin/git --version`**
2. **cannot serialize object of type `org.gradle.api.internal.project.DefaultProject`, a subtype of `org.gradle.api.Project`, as these are not supported with the configuration cache**

> The first error occurs when you start external processes without using `ExecOperations`. Gradle wants to know about external processes

Reviewer: >Gradle wants to know about external processes

... so it can decide whether or not it needs to rerun the configuration task as an input to the configuration cache has changed?

Author: I've added a sentence explaining this, let me know if it captures what you want to express here.

            int exitCode = process.waitFor();
            if (exitCode != 0) {
                throw new RuntimeException("zstd compression failed with exit code: " + exitCode);
            }
        } catch (IOException | InterruptedException e) {
            throw new RuntimeException("Failed to execute zstd compression", e);
        }
    }
}
```

### Step 1: Fixing "external process started"

- To safely do the external call, we need to use `ExecOperations` in `ZstdCompressor`
- To use `ExecOperations`, we can first make `ZstdCompressor` a Gradle-managed type, then inject `ExecOperations`
- To make `ZstdCompressor` a Gradle-managed type, we make it abstract. It now needs to be instantiated with `ObjectFactory::newInstance`
- To get an `ObjectFactory`, we inject it into `ZstdCompressTask`, which is conveniently already a Gradle-managed type

```java
abstract class ZstdCompressTask extends org.gradle.api.DefaultTask {
    @Input
    protected abstract Property<Path> getInputFile();

    @Input
    protected abstract Property<Path> getOutputFile();

    @Inject
    protected abstract ObjectFactory getObjectFactory();

    @TaskAction
    public void compress() {

>         ZstdCompressor compressor = getObjectFactory().newInstance(ZstdCompressor.class, 0);

Reviewer: The ideal would be to make this an `@Internal @Nested` but I wonder whether that's too many ideas to introduce here...

Author: We could, but I wanted to give the example of a regular stateful java class with non-static members and constructors with arguments. I think if people know how to deal with this case, then other simpler cases should be self-evident.

But yep @Nested is a neat trick people might appreciate. I've added a small snippet explaining the use of @Nested in the guide, let me know what you think of it.

    @Inject
    protected abstract ObjectFactory getObjectFactory();

    public addEmployee() {
        Employee employee = getObjectFactory().newInstance(Employee.class, "Kelvin");
        employees.add(employee);
    }
}
```

...so on and so forth...

```java
abstract class Startup {
    private int valuation;

    @Nested
    public abstract Team getTeam();
 
    public Startup(int valuation) {
        this.valuation = valuation; 
    }
}

```

> [!INFO]
> `@Nested` is an alternative to `ObjectFactory::newInstance` if your class has a nullary constructor, 

...until you reach the `Task`/`Project`/`Extension` that uses our class.

```java
abstract class UploadCompanyInformation extends DefaultTask {
    @Input
    protected abstract Property<URI> getUploadUri();

    @Inject
    protected abstract ObjectFactory getObjectFactory();

    @TaskAction
    public void compress() {
        Startup startup = getObjectFactory().newInstance(Startup.class, 1000000000000000);
        UploadUtils.upload(startup, getUploadUri());
    }
}
```

This way, you "propagate" the reach of the Gradle managed type system, removing boilerplate and making Configuration Cache adoption easy. 

> ### You can inject services into Gradle-managed types

Author: As discussed in person, lets put this in front of "### You can make any class into a Gradle-managed type"

<!-- PreviousNext:START -->
<table><tr>
  <td>Previous: <a href="testing-and-running-your-plugins-locally.md">Testing and running your plugins</a></td>
  <td align="center"><a href="../README.md#table-of-contents">Table of Contents</a></td>
  <td align="right">Next: <a href="diagnosing-build-performance.md">Diagnosing Build Performance</a></td>
</tr></table>
<!-- PreviousNext:END -->

# Adopting the Configuration Cache

The Gradle Configuration Cache is a powerful new feature. It reduces build latency by caching the task graph from the configuration phase.

Builds run with the Configuration Cache optionally in Gradle 8, by default in Gradle 9, and necessarily by Gradle 10. Furthermore, your build has to meet [strict requirements](https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:requirements) to be compatible. This guide provides practical steps to adopt the Configuration Cache.

> ## How does Configuration Caching work?

Reviewer 2: ```suggestion
## How does the Configuration Cache work?
```

Gradle builds are split into two phases — the configuration phase creates the task graph, while the execution phase runs it. The configuration phase takes various inputs and produces a complete task graph that defines what work needs to be done and in what order.

### Configuration inputs

1. Gradle environment
   - `GRADLE_USER_HOME`
   - Gradle Daemon JVM
2. Init scripts

> 3. buildSrc and included build logic build contents (build scripts, sources, and intermediate build outputs)

Reviewer 2: ```suggestion
3. `buildSrc` and included build logic build contents (build scripts, sources, and intermediate build outputs)
```

> 4. Build and Settings scripts, including included scripts (apply from: foo.gradle)

Reviewer 2: ```suggestion
4. Build and Settings scripts, including included scripts (`apply from: 'foo.gradle'`)
```

> 5. Gradle configuration files (Version Catalogs, dependency verification files, dependency lock files, gradle.properties files)

Reviewer 2: ```suggestion
5. Gradle configuration files (Version Catalogs, dependency verification files, dependency lock files, `gradle.properties` files)
```

6. Contents of files read at configuration time
7. File system state checked at configuration time (file presence, directory contents, etc.)

> 8. Custom ValueSource values obtained at configuration time (this also includes built-in providers, like providers.exec and providers.fileContents).

Reviewer 2: ```suggestion
8. Custom `ValueSource` values obtained at configuration time (this also includes built-in providers, like `ProviderFactory#exec` and `ProviderFactory#fileContents`.
```

9. System properties used during the configuration phase
10. Environment variables used during the configuration phase

### Configuration outputs

> The configuration phase output is a task [DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph), describing the target task and it's dependencies. Different tasks have different dependencies, and thus produce different task graphs. However, the target is always at the root of the DAG. Here are some of the task graphs of this repository:  

Reviewer 2: ```suggestion
The configuration phase output is a [directed acyclic graph (DAG)](https://en.wikipedia.org/wiki/Directed_acyclic_graph) of tasks, describing the target task and its dependencies. Different tasks have different dependencies, and thus produce different task graphs. However, the target task is always at the root of the DAG. Here are some of the task graphs of this repository:  
```

|         |    +--- :compileJava                               (org.gradle.api.tasks.compile.JavaCompile)
|         |    `--- :processResources                          (org.gradle.language.jvm.tasks.ProcessResources)
|         `--- :compileJava                                    (org.gradle.api.tasks.compile.JavaCompile)
`--- :check                                                    (org.gradle.api.DefaultTask)
   +--- :checkstyleMain                                      (org.gradle.api.plugins.quality.Checkstyle)
   |    +--- ...
   +--- :checkstyleTest                                      (org.gradle.api.plugins.quality.Checkstyle)
   |    +--- ...
   +--- :spotlessCheck                                       (org.gradle.api.DefaultTask)
   |    `--- ...
   +--- :test                                                (org.gradle.api.tasks.testing.Test)
   |    `--- ...
   ...
```

When you first run a gradle task, the task graph is serialized and stored on disk. Upon a subsequent run of the same task, If none of these inputs have changed, Gradle skips the configuration phase entirely, loads the task graph from disk, and goes straight to the execution phase.

> 

Reviewer 2: You probably want to move this section slightly above `Configuration Inputs`, the main motivation/reasoning is hidden underneath walls of code.

I also think you're underselling it a bit here. 

You want something along the lines of:
* running the configuration phase can be quite costly
* assuming it is deterministic given the inputs, the output (task graph) can be cached
* Rerun all the inputs again and if so, reuse without having to recalculate inputs.

I think you've got most of it already, just more emphasis on the configuration phase being costly (in and of itself).

Reviewer 2: There's also something about the verbiage which would be good to disambiguate.

It's not *just* when a task is run. It's more about the tasks requested i.e. what did you pass to `./gradlew <tasks>`. 
* It can be multiple tasks, not necessarily singular. We've been targeting the `classes` because that's the workflow we initially care about
* It can be a bit confusing to mention a task that's not in the requested task but is a dependant of one of the requested task. Especially when the log output is typically from the tasks requested.

Don't think you necessarily need to explain the distinction, but it would be worth being precise about it. e.g. mentioning what tasks requested means e.g. `./gradlew <tasks-requested...>` and talking about when you request the same set of tasks etc.

Author: > You probably want to move this section slightly above Configuration Inputs, the main motivation/reasoning is hidden underneath walls of code.

I put this under Configuration Inputs and Configuration Outputs, because they are prerequisites to understanding how CC works. However, I get your point, so I'll add a header to this core motivation section to make it more obvious. 

> just more emphasis on the configuration phase being costly (in and of itself)

Gotcha, I've updated the paragraph. 

> mentioning what tasks requested means e.g. ./gradlew <tasks-requested...> and talking about when you request the same set of tasks etc.

Gotcha, added the distinction to the paragraph.

> The configuration phase typically runs faster than execution since heavy work belongs in the latter. However, without the cache, configuration phase is rerun with every single Gradle run. If you run a unit test multiple times, configuration is repeated, reproducing the same task graph every time. When the task itself is light, configuration can take up a large fraction of the latency. Configuration caching solves this by storing and reusing configuration results between runs, eliminating redundant work and speeding up iteration cycles.

Reviewer 2: Feel like you're underselling it a bit here as well. Why is unit testing *such* a good example? because it requires a **tight feedback loop**. The task can be light, or medium, but it can still feel like it takes a while before anything "happens" i.e. paying for configuration.

Author: This comment clarifies what you mean! Added.

## Finding Configuration Cache problems

To find Configuration Cache problems:

> 1. Run `./gradlew build --configuration-cache`

Reviewer 2: I think `build` is way too large a scope, and probably not that useful as a starting point. It will likely be quite overwhelming and you probably won't see results for a long time!

It's probably good as a way to maybe capture everything as lots of tasks hook on to the `build` lifecycle task transitively, but it's not a given.

I think you probably want to mention starting off with a workflow you want to "speed up" e.g. compilation workflow typically invokes: `classes` and `testClasses`, so that's what you try to find problems with.

Author: Good idea!

2. Check the output for any problems. If none appear, great! Your build is compatible.
3. If not, fix the problems.
4. Run `./gradlew build --configuration-cache` again. In our experience, this reveals additional issues not seen in the first run.

The problems typically fall into three categories:

1. **external process started `/usr/bin/git --version`**
2. **cannot serialize object of type `org.gradle.api.internal.project.DefaultProject`, a subtype of `org.gradle.api.Project`, as these are not supported with the configuration cache**
3. **invocation of `Task.project` at execution time is unsupported**

> The first error occurs when you start external processes during the configuration phase. Generally, it is preferred to run external processes in tasks with properly declared inputs and outputs to avoid unnecessary work when the task is UP-TO-DATE. 

Reviewer 2: This is true, but there's also times where you do need to run things at configuration to decide when/what to configure a la palantir-java-format.

Author: True, but maybe that's a little bit too much detail for this guide?

> The second and third errors occur because Gradle tasks shouldn't take mutable types as input — `Gradle`, `Settings`, `Project`, `SourceSet`, or `Configuration`. Having these as inputs limits task concurrency (what if two tasks mutate a `Project` concurrently?) and prevents Gradle from serializing task inputs into the cache (they are too large to be serialized). 

Reviewer 2: I don't think size is necessarily an issue here. It's more that you can't reason about the state and reorder and parallelise things if you're reaching into other projects and using that to configure yourself. This is the reason why it limits task concurrency (*within* a project).

It also doesn't want you mutating the big mutable state at task execution time either, you could effectively change how things are and end up with a cache that's inconsistent with reality.

Author: Got it, removed the inaccurate claim that size was the issue.

To solve the second and third errors, you can

> - Declare the smallest "surface area" your task needs as task input, e.g. take in `Property<String>` for project version, instead of taking in `Project` and accessing `project.version`

Reviewer 2: ```suggestion
- Declare the smallest "surface area" your task needs as task input, e.g. take in `Property<String>` for the project version, instead of taking in `Project` and accessing `Project#version`
```

> - [Inject a service](https://docs.gradle.org/current/userguide/service_injection.html) which provides the information/operation you need

Reviewer 2: Let's just make these into subheadings:

### Most common categories of errors
// bla some preamble if it's useful

#### External process started e.g. `/usr/bin/git --version`
This occurs when you start external processes during the configuration phase. Generally, it is preferred to run external processes in tasks with properly declared inputs and outputs to avoid unnecessary work when the task is UP-TO-DATE.

etc.

Author: Yep!

## Solving Configuration Cache problems

Let’s walk through some examples of fixing Configuration Cache problems

> [!TIP]

> > If you have a huge Gradle projects with many tasks, you can adopt the Configuration Cache incrementally. First, resolve all configuration phase issues, as incremental adoption only works for the execution phase. Then, apply the [gradle-incremental-configuration-cache](https://github.com/palantir/gradle-incremental-configuration-cache) plugin. Gradually add tasks to the allow list as they become compatible.

Reviewer 2: This seems somewhat out of place:

> First, resolve all configuration phase issues

This is what this whole section is about? Also have we discussed what the difference between configuration phase cc issues and execution phase cc issues are? There's lots to unpack in this tip that is not obvious to me.

This tip probably deserves its own section, at the bottom

Author: > This is what this whole section is about

Hmm, this section talks about both configuration-phase (e.g. external process started) and execution-phase (e.g. calling Task#project in @TaskAction) issues. 

> have we discussed what the difference between configuration phase cc issues and execution phase cc issues are?

I've added a clarification that this refers to problems like `external process started`.

Also, this guide is intended to be standalone, independent of `gradle-incremental-configuration-cache`. I don't think having a separate section for the plugin respects this.

### Fixing "external process started"

#### Before: 
- `MyPlugin` is already a [Gradle managed type](managed-types-and-properties.md)

> - `MyPlugin::apply` starts an external process in the configuration phase, causing the configuration cache to fail

Reviewer 2: ```suggestion
- `MyPlugin#apply` starts an external process in the configuration phase, causing the configuration cache to fail
```

Reviewer 2: explain in a bit more detail what is actually happening here. i.e. `getLatestGitTag` starts an external process calling out to `git`.

Reviewer 2: make the example a bit more useful, *use* the result somewhere.

Author: > explain in a bit more detail what is actually happening here. i.e. getLatestGitTag starts an external process calling out to git.

Gotcha

> make the example a bit more useful, use the result somewhere.

Hmm, I think adding any more code to the example will make it too complicated. Let's keep it simple, and just use a generic `// Do something...` block?

```java
abstract class MyPlugin implements Plugin<Project> {
   @Override
   public final void apply(Project project) {
      String gitTag = getLatestGitTag(project);
      if (gitTag.equals("develop")) {
         // Do something...
      }
   }

   private static String getLatestGitTag(Project project) {
      try {
         Process process = new ProcessBuilder("git", "describe", "--tags", "--abbrev=0")
                 .directory(project.getProjectDir())
                 .start();
         try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
            String tag = reader.readLine();
            process.waitFor();
            return tag.strip();
         }
      } catch (Exception e) {
         throw new RuntimeException(e);
      }
   }
}
```

#### After

- To safely do an external call during the configuration phase, we can use the `ExecOperations` service.

> - To use `ExecOperations`, we can inject it into `MyPlugin`. 

Reviewer 2: I don't think you want to use `ExecOperations`, that is largely for avoiding issues *within* a Task action i.e. using `project` at execution time. You want to use `ProviderFactory#exec` as that then gives you back a provider that's lazy. You can then pass that around and you may not even execute it! => if it's not on your task graph, you have less things to run to compute your cache key.

If you compare it to some of the work you were doing, we shied away from it for that same reason.

If you have a look at https://docs.gradle.org/current/userguide/configuration_cache.html#config_cache:requirements:external_processes they make a similar argument i.e. use `providers.exec` or have your own value source that does what you want and gives you a provider back.

Reviewer 2: It also means that you don't actually end up talking about the pitfall of the `ProviderFactory#exec` where you need to keep a reference to the returned provider.

Author: > I don't think you want to use ExecOperations, that is largely for avoiding issues within a Task action i.e. using project at execution time. You want to use ProviderFactory#exec as that then gives you back a provider that's lazy. You can then pass that around and you may not even execute it! 

Thanks for catching this. 

I think i was confused by the wording in Gradle docs (filed [an issue here](https://github.com/gradle/gradle/issues/34293))

But yep, your point is to prefer `ProviderFactory#exec` over other methods like `Project#exec`, `ExecOperations#exec`. Because `ProviderFactory#exec` returns a provider which may or may not be executed, but the other methods execute straight away. 

> It also means that you don't actually end up talking about the pitfall of the ProviderFactory#exec where you need to keep a reference to the returned provider.

Added a [!TIP] box for this!

```java
abstract class MyPlugin implements Plugin<Project> {
   @Inject
   protected abstract ExecOperations getExecOperations();

   @Override
   public final void apply(Project project) {
      String gitTag = getLatestGitTag(project);
      if (gitTag.equals("develop")) {
         // Do something...
      }
   }

   private String getLatestGitTag(Project project) {
      try {
         OutputStream output = new ByteArrayOutputStream();
         getExecOperations().exec(execSpec -> {
            execSpec.workingDir(project.getProjectDir());
            execSpec.setCommandLine("git", "describe", "--tags", "--abbrev=0");
            execSpec.setStandardOutput(output);
         });
         return output.toString().strip();
      } catch (Exception e) {
         throw new RuntimeException(e);
      }
   }
}
```

### Fixing "invocation of `Task.project` at execution time is unsupported"

#### Before
- `MyTask` is already a [Gradle managed type](managed-types-and-properties.md)

> - `MyTask` is calling `getProject()` at build time, causing the configuration cache to fail: **invocation of `Task.project` at execution time is unsupported**

Reviewer 2: ```suggestion
- `MyTask` is calling `getProject()` at build time, causing the configuration cache to fail with 

    invocation of `Task.project` at execution time is unsupported
```

let's wrap it in a code block

Author: Gotcha.

```java
abstract class MyTask extends DefaultTask {
   @InputFile
   protected abstract FileProperty getMyFiles();

   @TaskAction
   public void delete() {
      boolean deleted = getProject().delete(getMyFiles().getFiles());
      // Do something with `deleted`...
      Path projectDir = getProject().getProjectDir().toPath();
      // Do something with `projectDir`...
   }
}

```

#### After

> - To get the project dir, we can use the `ProjectLayout` service.

Reviewer 2: ```suggestion
- To get the project directory, we can use the `ProjectLayout` service.
```

I don't really like it for code, but understandable, but let's use full names for things when writing docs/prose please.

Author: Gotcha

- To do file system operations like delete, we can use the `FileSystemOperations` service

> - We can inject these services into `MyTask`

Reviewer 2: make a note about the visibility requirements, needing to use `@Inject` and prefixing with `get`. It's a common enough pitfall and the error message sucks when you don't get the correct invocation

Author: Yep!

```java
abstract class MyTask extends DefaultTask {
   @InputFile
   protected abstract FileProperty getMyFiles();

   @Inject
   protected abstract ProjectLayout getProjectLayout();

   @Inject
   protected abstract FileSystemOperations getFileSystemOperations();

   @TaskAction
   public void delete() {
      WorkResult result = getFileSystemOperations().delete(spec -> spec.delete(getMyFiles().getFiles()));
      // Do something with `result`...
      Path projectDir = getProjectLayout().getProjectDirectory().getAsFile().toPath();
      // Do something with `projectDir`...
   }
}
```

## Two strategies to write Configuration Cache friendly Gradle

To summarize the two strategies we used to solve Configuration Cache issues:

### You can inject services into Gradle-managed types

> Do you need the project directory? Inject [`ProjectLayout`](https://docs.gradle.org/current/javadoc/org/gradle/api/file/ProjectLayout.html). Do you need to run a bash command? Inject [`ExecOperations`](https://docs.gradle.org/current/javadoc/org/gradle/process/ExecOperations.html). Gradle offers a [list](https://docs.gradle.org/current/userguide/service_injection.html) of injectable services which will cover 99% of your use cases. However, that list is incomplete — in reality,  almost [anything in Gradle source](https://github.com/search?q=repo%3Agradle%2Fgradle%20%40ServiceScope&type=code) annotated with [`@ServiceScope`](https://github.com/gradle/gradle/blob/196bb409d47f5b6e39d62edd39be939f7606a5cc/platforms/core-runtime/stdlib-java-extensions/src/main/java/org/gradle/internal/service/scopes/ServiceScope.java#L43) can be injected into a Gradle managed type.

Reviewer 2: * as stated above, the bash command is context dependent, are you in a task action? are you in a plugin code?

Author: Yep, let's specify that we only use `ExecOperations` to run bash commands at configuration time.

### You can make any class into a Gradle-managed type

> A standard class...

Reviewer 2: You should explain the benefit to doing this, it appears random here.

You haven't actually made any Gradle-managed types in the examples given

Author: Let's make some gradle-managed types in the examples!

    private String name;

    @Inject
    public Employee(String name) {
        this.name = name;
    }
}
```

Now, we can instantiate it with `ObjectFactory`.

```java
Employee employee = objectFactory.newInstance(Employee.class, "Kelvin");
```

Hold on, what if the class using `Employee` doesn't have access to an `ObjectFactory`? 

```java
class Team {
    private ArrayList<Employee> employees = new ArrayList<>();

    public addEmployee() {
        Employee employee = new Employee("Kelvin");
        employees.add(employee);
    }
}
```

In that case, you can turn *it* a Gradle managed type as well...

```java
abstract class Team {
    private ArrayList<Employee> employees = new ArrayList<>();

    @Inject
    protected abstract ObjectFactory getObjectFactory();

    public addEmployee() {
        Employee employee = getObjectFactory().newInstance(Employee.class, "Kelvin");
        employees.add(employee);
    }
}
```

...so on and so forth...

```java
abstract class Deployment {
    private String customerName;

    @Nested
    public abstract Team getTeam();
 
    @Inject
    public Deployment(String customerName) {
        this.customerName = customerName; 
    }
}

```

> > [!TIP]

Reviewer 2: I think you're underselling this!

This probably deserves its own section and/or have most of the examples work via `@Nested`. This is the end goal of where you will probably want to end up in most cases.

Author: Yep, let's add a third example, where we explicitly show the benefits of @Nested

## Finding Configuration Cache problems

To find Configuration Cache problems:

1. Run `./gradlew build --configuration-cache`
2. Check the output for any problems. If none appear, great! Your build is compatible.
3. If not, fix the problems.
4. Run `./gradlew build --configuration-cache` again. In our experience, this reveals additional issues not seen in the first run.

The problems typically fall into three categories:

1. **external process started `/usr/bin/git --version`**
2. **cannot serialize object of type `org.gradle.api.internal.project.DefaultProject`, a subtype of `org.gradle.api.Project`, as these are not supported with the configuration cache**
3. **invocation of `Task.project` at execution time is unsupported**

> The first error occurs when you start external processes during the configuration phase. Generally, , it is preferred to run external processes in tasks with properly declared inputs and outputs to avoid unnecessary work when the task is UP-TO-DATE. 

Reviewer: > , ,

> without using `ExecOperations`. Gradle needs to be aware of all external processes so it can determine if any changes require the configuration cache to be re-run.

Reviewer: > without using `ExecOperations`

> The second and third errors occur because Gradle tasks shouldn't take mutable types as input — Gradle, Settings, Project, SourceSet, or Configuration. Having these as inputs limits task concurrency (what if two tasks mutate a `Project` concurrently?) and prevents Gradle from serializing task inputs into the cache (they are too large to be serialized). 

Reviewer: Generally, I'd wrap code references like Gradle, Settings etc in \` so they appear like `Gradle`, `Settings` etc

Author: Yep I missed these!

To solve the second and third errors, you can
- Declare the smallest "surface area" your task needs as task input, e.g. take in `Property<String>` for project version, instead of taking in `Project` and accessing `project.version`

> - [Inject a service](https://docs.gradle.org/current/userguide/service_injection.html) which provides the information/operation you need

Reviewer: I would give `FileOperations` and `ExecOperations` as examples.

Author: Yep I've updated this section to have two examples, a FileSystemOperations one and an ExecOperations one

## Solving Configuration Cache problems

Let’s walk through making a task compatible with the Configuration Cache.

> [!TIP]
> If you have a huge Gradle projects with many tasks, you can adopt the Configuration Cache incrementally. First, resolve all configuration phase issues, as incremental adoption only works for the execution phase. Then, apply the [gradle-incremental-configuration-cache](https://github.com/palantir/gradle-incremental-configuration-cache) plugin. Gradually add tasks to the allow list as they become compatible.

### Before:

- `CompanyReportPlugin` and `GenerateReportTask` are already [Gradle managed types](managed-types-and-properties.md)
- `ReportGenerator` is a plain old Java object.
- `CompanyReportPlugin::apply` starts an external process in the configuration phase, causing the configuration cache to fail

> - Furthermore, `GenerateReportTask` is calling `getProject()` at build time, causing another configuration cache failure: **invocation of `Task.project` at execution time is unsupported**

Reviewer: No need for "Furthermore" in bullet points.

```java
abstract class CompanyReportPlugin implements Plugin<Project> {
   @Override
   public final void apply(Project project) {
      String gitTag = getLatestGitTag(project);
      if (gitTag.equals("develop")) {
         project.getTasks().register("generateReport", GenerateReportTask.class);
      }
   }

   private static String getLatestGitTag(Project project) {
      try {
         Process process = new ProcessBuilder("git", "describe", "--tags", "--abbrev=0")
                 .directory(project.getProjectDir())
                 .start();
         try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
            String tag = reader.readLine();
            process.waitFor();
            return tag.strip();
         }
      } catch (Exception e) {
         throw new RuntimeException(e);
      }
   }
}

abstract class GenerateReportTask extends DefaultTask {
   @TaskAction
   public void upload() {

>       SassyReporter generator = new ReportGenerator(1000);

Reviewer: What is `SassyReporter`? Should it be just `ReportGenerator`?

Author: That was the original name hahaha. Let's remove it
